antsibull/docs_parsing/routing.py

```diff
@@ -0,0 +1,309 @@
+# coding: utf-8
+# Author: Felix Fontein <felix@fontein.de>
+# License: GPLv3+
+# Copyright: Ansible Project, 2021
+"""
+Functions for parsing and interpreting collection metadata.
+"""
+
+from collections import defaultdict
+
+import datetime
+import os
+import typing as t
+
+import asyncio
+import asyncio_pool
+
+from ansible.constants import DOCUMENTABLE_PLUGINS
+
+from .. import app_context
+from .. import yaml
+from ..utils.get_pkg_data import get_antsibull_data
+from . import AnsibleCollectionMetadata
+from .fqcn import get_fqcn_parts
+
+
+# A nested structure as follows:
+#       plugin_type:
+#           plugin_name:  # FQCN
+#               tombstone: t.Optional[{tombstone record}]
+#               deprecation: t.Optional[{deprecation record}]
+#               redirect: t.Optional[str]
+#               redirect_is_symlink: t.Optional[bool]
+CollectionRoutingT = t.Mapping[str, t.Mapping[str, t.Mapping[str, t.Any]]]
+
+
+COLLECTIONS_WITH_FLATMAPPING = (
+    'community.general',
+    'community.network',
+)
+
+
+def add_symlink(collection_name: str, src_components: t.List[str], dest_components: t.List[str],
+                plugin_type_routing: t.Dict[str, t.Dict[str, t.Any]]) -> None:
```

Comment:
As much as possible, I like to create functions rather than procedures that operate by side effect.  I think this one definitely can return its value rather than modifying `plugin_type_routing`.  I have a feeling that `add_symlinks()` can also return its value but I'm not as sure there.  returning the value would mean moving some logic to the higher level and I haven't quite decided if that would be a good thing or not.

Reasons for liking functions:
* Side effects are harder to keep in mind than return values.
* Unit testing is more straightforward for functions than procedures which operate by side effect.